[TOC]

![img](image/Fhcw4wifVKoZoDpVK2CJqyyuEycU.png)

# Hydrolysis of nucleic acid（核酸水解）

+   The nucleic acid can be broken down into the free bases, phosphoric acid, and(deoxy) ribose during hydrolysis.![image-20211021141834310](image/image-20211021141834310.png)

+   The glycosidic bonds(糖苷键) and phosphodiester bond(磷酸二酯键) in nucleic acid can be broken down by acids(酸), alkali(碱), enzymes(酶).

+   Effect of acid(酸)
    +   RNA and DNA are insoluble in cold acid. DNA is more sensitive to acid hydrolysis than RNA.
    +   In a neutral pH range, pH 5 to 9, DNA molecules are quite stable.
        However, if the pH becomes too acidic or alkaline,DNA molecules are prone to destabilization.
    +   At pH 5 or lower, DNA is liable to depurination(去嘌呤化) (i.e. the loss of purine bases from DNA).The phosphodiester bonds of DNA break which causes the base pairs of DNA to break off.
        +   Depurination in DNA is a chemical reaction of purine deoxyribonucleosides，deoxyadenosine and deoxyguanosine，in which the β-N-glycosidic bond（糖苷键）is hydrolytically cleaved releasing a nucleic base，adenine or guanine.
            ![image-20211021142054306](image/image-20211021142054306.png)
        +   The electrophilic attack of acid on N7 gives rise to depurination by hydrolysis
            ![image-20211021142138615](image/image-20211021142138615.png)
    +   Treatment with 1N HCI at 60 ℃ for 10 minutes will not hydrolyze RNA but can depurinate DNA. This is used in Feulgen staining procedure and makes it DNA specific staining.
    +   At pH<1, however, both DNA and RNA break down into the free bases, phosphoric acid, and(deoxy) ribose.
+   Effect of alkali（碱）
    +   At pH 9 or higher，negatively-charged ions remove hydrogen ions from the base pairs of DNA，thereby breaking the hydrogen bonds between and causing the DNA strands to denature. While less damaging in comparison to acidic pH levels，high alkaline pH levels can still result in depurination.
    +   RNA is hydrolyzed rapidly under alkaline conditions，but DNA is not. RNA can be easily hydrolyzed at alkali pH as it contains 2'-OH
        ![image-20211021142331745](image/image-20211021142331745.png)
    +   Schematic of RNA degradation under extreme alkaline(碱性的)pH conditions
        +   ![image-20211021142423201](image/image-20211021142423201.png)
+   Effect of enzymes(酶)
    +   Nucleases(核酸酶) are a class of enzymes called hydrolases(水解酶)
    +   Break nucleic acids into smaller subunits called nucleotides by hydrolyzing the phosphodiester bonds(磷酸二酯键) present between them
        ![image-20211021142501070](image/image-20211021142501070.png)
    +   Classification of Nucleases 
    +   Based on substrate preference,they can be grouped into DNases(脱氧核糖核酸酶) and RNases(核糖核酸酶).
    +   Based on their site of action along the nucleic acid strand,nucleases can be grouped into exonucleases(外切核酸酶) and endonucleases(内切核酸酶)![image-20211021142803663](image/image-20211021142803663.png)
    +   Restriction enzyme(Restriction endonuclease 限制性内切酶),is an enzyme that cleaves DNA into fragments at or near specific recognition sites within molecules known as **restriction sites**.
    +   Production of blunt ends and sticky ends.![image-20211021142843060](image/image-20211021142843060.png)

## Restriction endonucleases are very useful in recombinant DNA technology

![image-20211021143321541](image/image-20211021143321541.png)

+   Recombinant DNA technology has been exploited in order to provide selective improvements in various specialties that include crop agriculture, pharmaceutics, gene therapy, vaccine design and bioremediation.
+   The most common application of recombinant DNA is in basic research, identify, map and sequence genes, and to determine their function

## The CRISPR/Cas9 system for genomic engineering

**Clustered Regularly Interspaced Short Palindromic Repeats**

![image-20211021143450228](image/image-20211021143450228.png)

![image-20211021143514708](image/image-20211021143514708.png)

![image-20211021143536747](image/image-20211021143536747.png)

![image-20211021143614376](image/image-20211021143614376.png)

![image-20211021143635453](image/image-20211021143635453.png)

![image-20211021145147455](image/image-20211021145147455.png)

![image-20211021145204380](image/image-20211021145204380.png)

>CRISPR是细菌染色体的一部分

+   外源基因的捕获

![image-20211021145438189](image/image-20211021145438189.png)

![image-20211021145420333](image/image-20211021145420333.png)

![image-20211021145453575](image/image-20211021145453575.png)

+   crRNA 的形成

![image-20211021145521755](image/image-20211021145521755.png)

![image-20211021145536646](image/image-20211021145536646.png)

![image-20211021145551493](image/image-20211021145551493.png)

+   DSB 染色体双链断裂
    +   双链断裂之后一般有两种方式修复 DSB
        +   NHEJ 非同源末端链接
            +   此过程不需要同源序列当模板，容易造成碱基的随机插入和缺失，导致基因功能性缺失，利用这个特点可以实现基因的敲除
            +   ![image-20211021145654202](image/image-20211021145654202.png)
        +   HR 同源重组
            +   这种方式需要同源序列当模板，能够实现真正意义上的修复
                这个过程只发生在细胞周期的S或G2期
            +   ![image-20211021145727653](image/image-20211021145727653.png)



## Acid-base properties of the nucleic acid (核酸酸碱性质)

+   Nucleic acid bases(碱基), nucleosides(核苷) and nucleotides(核苷酸)can all be dissociated. Under certain conditions,zwitterion(兼性离子) can be formed,which is an amphoteric electrolyte(两性电解质) with an isoelectric point(等电点).

![image-20211021143732058](image/image-20211021143732058.png)

## Ultraviolet absorption properties of the nucleic acid（核酸紫外吸收性质）

![image-20211021143902377](image/image-20211021143902377.png)

+   Nucleic acids absorb ultraviolet(UV) light due to **conjugated double bonds**(共轭双键) in their purine and pyrimidine rings; the sugar-phosphate backbone does not contribute to absorption.

+   The wavelength of **maximum absorption for both DNA and RNA is 260nm.**

![image-20211021143946921](image/image-20211021143946921.png)

+   The absorption properties of DNA can be used for detection,quantification and assessment of purity.
+   The standard nucleic acid quantitation method is UV spectrophotometry (紫外分光光度计法)

![image-20211021144254893](image/image-20211021144254893.png)

+   Measures DNA, RNA(A260) and Proteins(A280) concentrations and sample purity (A260:A280).
    +   ![image-20211021144600025](image/image-20211021144600025.png)
    +   A ratio 260/280 of~1.8 is generally accepted as"pure"for DNA; 
    +   A ratio of ~2.0 is generally accepted as "pure"for RNA.
    +   Similarly, absorbance at 230 nm is accepted as being the result of other contamination; 
    +   Expected 260/230 values are commonly in the range of 2.0-2.2.

## Nucleic acid denaturation and renaturation (核酸的变性和复性)

+   Nucleic acid denaturation (核酸的变性) occurs when hydrogen bonding between nucleotides is disrupted, and results in the separation of previously annealed strands.
+   Denatured DNA can be obtained in high temperatures, high concentration of denaturant (urea, methylmercuric hydroxide, etc.) or at extreme pH values.
+   For example,denaturation(变性) of DNA due to high temperatures results in the disruption of Watson and Crick base pairs and the separation of the double stranded helix into two single strands.

![image-20211021150137084](image/image-20211021150137084.png)

+   DNA denaturation can be followed spectrophotometrically by measuring the absorption of UV light at 260 nm 
+   In its native state,DNA absorbs less UV light than the separate polynucleotide chains,a phenomenon that is called hypochromic effect(减色效应).
+   If a DNA dispersion is slowly heated and its UV light absorption isfollowwed,the temperature/absorbency relationship is an indicator of DNA denaturation

![image-20211021150209596](image/image-20211021150209596.png)

+   As the temperature augments，an increase of absorption(hyperchromic effect(增色效应) ) occurs.
+   When a given temperature is reached，the optical density does not further increase，which shows that DNA chains are completely separated and the molecules are maximally denatured.

![image-20211021150306626](image/image-20211021150306626.png)

#### Melting Temperature($T_m$)(熔解温度)

![image-20211021150434893](image/image-20211021150434893.png)

The temperature at which half of the DNA is denatured(corresponding to the midpoint or in flection of the curve)is known as the DNA melting temperature($T_m$)

+   The Tm is characteristic for each DNA under defined conditions of pH and salt concentration, it ranges from 80 to 100℃ for DNA isolated from different organisms.
+   Determining the value of Tm is useful to estimate the base composition of DNA.
    $A=T, G\equiv C$

#### Correlations of Melting Temperature ($T_m$) with G + C Content

![image-20211021150610705](image/image-20211021150610705.png)

AT base pairs have two hydrogen bonds and GC pairs have three.
Therefore, there is more energy in the GC bond. Consequently, the higher the GC content of a molecule, the higher the melting temperature (more energy is necessary to make the helix-coil transition)

>   pH, 序列, 离子浓度

### The DNA renaturaion (复性)

The DNA denaturation process is reversible under controlled conditions of pH and ionic strength. If the temperature is slowly decreased in the solution where the DNA had been denatured, the DNA chains will spontaneously reanneal and the original double helix structure is restored.

![image-20211021150906119](image/image-20211021150906119.png)

If a hot solution of denatured DNA is **cooled slowly**, the hydrogen bonds between complementary bases and the **original dsDNA** are restored-the process is called **annealing**(退火)

If a hot solution is **cooled rapidly**，hydrogen bonds between bases are formed in a random way. Various either **intramolecular or intermolecular aggregates** are formed.

![image-20211021150937321](image/image-20211021150937321.png)

---

+   This process of DNA renaturation(复性) is also a kinetic one.
+   The reaction is a second order rate, dependent upon the concentration of DNA.
+   The **renaturation rate** of DNA is an excellent indicator of the **sequence complexity and the size of the DNA.**

### Renaturation rate and DNA sequence complexity -$C_0t$ Curves

Second-order rate equation for two complementary strands coming together is given by the rate of decrease in C
$$
\frac{\mathrm{d}C}{\mathrm{d}t}=-KC^2
$$

>   C is the concentration of single-stranded DNA at time t;
>   K is the second-order rate constant.

$$
\frac{\mathrm{d}C}{\mathrm{d}t}=-KC^2\Longrightarrow\frac{C}{C_0}=\frac{1}{1+KC_0t}
$$

>   $C_0$, Starting with a concentration of completely denatured DNA at $t=0$

Then the time for half of the DNA to renature(when $\frac{C}{C_0}=0.5$), according to the second order rate equation, is defined as $t=t_{\frac{1}{2}}$. Then
$$
C_0t=\frac{1}{K}
$$


If you plot the amount of the DNA that has renatured(reassociated)against the $C_0t$, you get curves such as 

$C_0t$ Curves

![image-20211021151641224](image/image-20211021151641224.png)

+   The shape of a $C_0t$ Curves for a given species is a function of two factors:
    +   The size or complexity of the genome
    +   The amount of repetitive DNA within the genome

## Nucleic acid hybridization(核酸的分子杂交)

+   Nucleic acid hybridization(核酸的分子杂交) is a fundamental tool in molecular genetics which takes advantage of the ability of individual single-stranded nucleic acid molecule to form double stranded molecules (that is,to hybridize to each other)

+   Standard nucleic acid hybridization assays 
    +   A labeled nucleic acid-a probe-to identifyrelated DNA or RNA molecules; 
    +   Complex mixture of unlabeled nucleic acid molecules - the target; 
    +   Base complementarity with a high degree of similarity between the probe and the target.
+   Southern blotting(Southern 印迹法) is a method used in molecular biology for detection of a specific DNA sequence in DNA samples.

![image-20211021152123257](image/image-20211021152123257.png)

This method is developed by British biologist Edwin Southern

+   Northern blotting(Northern 印迹法)or RNA blot,is a technique used in molecular biology research to study gene expression by detection of RNA(or isolated mRNA)in a sample.

![image-20211021153407776](image/image-20211021153407776.png)

# Isolation and purification of nucleic acids(核酸的分离纯化)

+   Basic nucleic acid (NA) preparation steps used in most isolation procedures.(Not all steps are used in all methods.)

![image-20211021153452711](image/image-20211021153452711.png)

+   Separation of Nucleic acids:
    +   By ultracentrifugation(超速离心)
    +   By column chromatography(柱层析)
    +   By gel electrophoresis(凝胶电泳)

## Density gradient centrifugation(密度梯度离心)

+   On the basis of their buoyant density(浮力密度),independently of their size or shape.
+   The sample is usually either layered on top of,or dispersed within,a steep density gradient that contains a very high concentration of sucrose(蔗糖)or cesium chloride(氯化铯).
+   Each subcellular component will move up or down when centrifuged until it reaches a position where its density matches its surroundings and then will move no further.
+   Series of distinct bands will eventually be produced,with those nearest the bottom of the tube containing the components of highest buoyant density.

![image-20211021153710480](image/image-20211021153710480.png)

+   Separation of Nucleic acid in sucrose(蔗糖) gradient

![image-20211021153931194](image/image-20211021153931194.png)

---

+   Separation of Nucleic acid in $CsCI$ (cesium chloride)(氯化铯) gradient

![image-20211021154004412](image/image-20211021154004412.png)

## Column Chromatography 柱层析

+   Several methods have been applied to the fractionation of nucleic acids including
    +   HPLC anion-exchange chromatography
    +   Methylated albumin-kieselguhr columns
    +   Silica column chromatography
    +   Cellulose powder(spin-)column chromatography
    +   Hydroxyapatite column

### Viral nucleic acid fractionation using tandem hydroxyapatite and cellulose column chromatography(羟磷灰石和纤维素柱层析法)

![image-20211023225201481](image/image-20211023225201481.png)

![image-20211023225221312](image/image-20211023225221312.png)

## Gel Electrophoresis 凝胶电泳

+   Physical method of separation of molecules based on their charge and length
    +   Agarose gel electrophoresis(琼脂糖凝胶电泳)
    +   Pulsed field gel electrophoresis(PFGE)(脉冲电场凝胶电泳)
    +   Polyacrylamide gel electrophoresis(PAGE)(聚丙烯酰胺凝胶电泳)

Electrophoretic band of DNA/RNA can be visualized under UV light (UV-transilluminator or Gel-Doc)at specific wavelength.

Visualizations achieved by different strategies. Most commonly by intercalating dyes(ethidium bromide(溴化乙锭),etc.).

###  DNA stained with ethidium bromide(溴化乙锭)

![image-20211023225411612](image/image-20211023225411612.png)

When exposed to ultraviolet light,it will fluoresce with an orange color, intensifying almost 20-fold after binding to DNA

### Agarose Gel Electrophoresis(琼脂糖凝胶电泳)

![image-20211023225457276](image/image-20211023225457276.png)

+   Agarose gel electrophoresis is a method to separate DNA, or RNA molecules by size. 
+   This is achieved by moving negatively charged nucleic acid molecules through an agarose matrix with an electric field(electrophoresis).
+   The pore size is determined by adjusting the concentration of agarose in a gel(normally in the rank of 0.4-4%)
+   Shorter molecules migrate farther than longer ones.

### Pulsed Field Gel Electrophoresis(PFGE)(脉冲电场凝胶电泳)

+   PFGE allows investigators to separate much larger pieces of DNA than conventional agarose gel electrophoresis.

+   In conventional gels, the current is applied in a single direction (from top to bottom). But in PFGE, the direction of the current is altered at a regular interval.

![image-20211023225732049](image/image-20211023225732049.png)

![image-20211023225755461](image/image-20211023225755461.png)

## DNA extraction and purification (DNA的提取和纯化)

+   There are several basic steps in a DNA extraction, the details of which may vary depending on the type of sample and any substances that may interfere with the extraction and subsequent analysis.
    +   Breaking the cells open to expose the DNA
    +   Removing the membrane lipids through the use of an appropriate detergent
    +   Removal of proteins by adding an appropriate protease
    +   Removal of contaminants
    +   Removal of RNA with the use of an RNase Precipitating the DNA with alcohol

![image-20211023230225831](image/image-20211023230225831.png)

+   Liquid Phase Organic Extraction
    +   Phenol: Chloroform: Isoamyl alcohol(50:49:1)
        +   Lysed samples mixed with above; two layers are formed
        +   Proteins remain at interface
        +   DNA is removed with top aqueous layer
        +   DNA is precipitated with alcohol and rehydrated.
    +   Disadvantages
        +   Slow, labor-intensive, toxic(phenol, chloroform)
        +   Fume hood required, disposal of hazardous materials required

+   Nonorganic Salt Precipitation 
    +   Cell membranes are lysed and proteins are denatured by detergent(such as sodium dodecyl sulfate(SDS)(十二烷基硫酸钠)).
        +   RNA is removed with RNase.
        +   Proteins are precipitated with salt solution.
        +   DNA is precipitated with alcohol and rehydrated.
    +   Advantages:
        +   Fast and easy method 
        +   Uses nontoxic materials,no fume hood required, no hazardous materials disposal issues
        +   Produces high-quality DNA

## RNA extraction and purification(RNA的提取和纯化)

+   There are several basic steps in a RNA extraction, the details of which may vary depending on the type of sampleand any substances that may interfere with the extraction and subsequent analysis.
    +   Breaking the cells open to expose the RNA(add RNase inhibitors)
    +   Removing the membrane lipids through the use of an appropriate detergent
    +   Removal of proteins by adding an appropriate protease
    +   Removal of contaminants
    +   Removal of DNA with the use of an DNase
    +   Precipitating the RNA.

---

+   RNA is not a stable molecule
    +   It is easily degraded by RNase enzymes.
+   Use sterile, disposable and RNase free plastic ware (tubes, filter tips)
+   Always wear gloves and work in a hood whenever possible/practical.
+   Treat liquids with diethyl pyrocarbonate(DEPC).
+   It is important to note that RNA is less stable as compared to DNA so one should only isolate it prior to use.

---

### Three major techniques extensively used for RNA extraction 

+   Organic extraction,such as Phenol(苯酚)/Guanidine Isothiocyanate(GITC)(异硫氰酸服)-based solutions
+   Silica-membrane based spin column technology
+   Paramagnetic particle technology.

####  Guanidinium(胍盐)-based Organic Isolation

+   Phenol(苯酚)/guanidinium(胍盐)solution disrupts cells,solubilizes cell components,but maintains integrity of RNA.
+   Add chloroform(氯仿),mix,and centrifuge.
+   Proteins/DNA remain at interface.
+   RNA is removed with aqueous top layer.
+   RNA is precipitated with alcohol and rehydrated.
+   Advantage: faster than $CsCI$ method 
+   Disadvantages: fume hood required, hazardous waste disposal issues

#### Cesium Chloride Gradient(氯化绝梯度离心)

Used mainly to get clean RNA for Northern blots 

+   Homogenize cells in guanidinium isothiocyanate and $β$-mercaptoethanol solution
+   Add to $CsCI$ gradient and centrifuge for 12-20 hours. RNA will be at the bottom of tube
+   Re-dissolve in TE/SDS buffer
+   Precipitate RNA with salt and ethanol, then rehydrate
+   Advantage: high quality materials disposal issues
+   Disadvantages: extremely time-consuming, hazardous materials disposal issues.

# Nucleic acid sequencing(核酸序列测定)

## History of sequencing technology

![image-20211023231154062](image/image-20211023231154062.png)

## DNA sequencing

### First generation sequencing(Sanger method)

#### The classical chain-termination method for DNA sequencing

![image-20211023231234562](image/image-20211023231234562.png)

### Second generation sequencing(massively parallel)

![image-20211023231253146](image/image-20211023231253146.png)

### Third generation sequencing(Nanopore sequencing(新型纳米孔测序法))

Nanopore sequencing is used in the sequencing of biopolymers, specifically, polynucleotides in the form of DNA or RNA.

![image-20211023231328132](image/image-20211023231328132.png)

## DNA sequencing applications

![image-20211023231343595](image/image-20211023231343595.png)

![image-20211023231349964](image/image-20211023231349964.png)

![image-20211023231354699](image/image-20211023231354699.png)

## RNA sequencing

An overview of library preparation methods for different RNA-sequencing(RNA-seq) methods

![image-20211023231420586](image/image-20211023231420586.png)

### The three main sequencing technologies for RNA-seq

![image-20211023231440458](image/image-20211023231440458.png)

### Comparison of short-read, long-read and direct RNA-seq analysis.

![image-20211023231452757](image/image-20211023231452757.png)

### The single-cell RNA sequencing(RNA-seq)workflow

![image-20211023231509586](image/image-20211023231509586.png)

### The spatialomics workflow

![image-20211023231519029](image/image-20211023231519029.png)

